site/src/Map.jsx

```diff
@@ -1,135 +1,313 @@
-import PropTypes from 'prop-types';
-import './InspectorPanel.css';
-import './CustomControls.css';
-import ThemeSelector from './ThemeSelector';
-
-const MAP_STYLE = 'https://basemaps.cartocdn.com/gl/positron-gl-style/style.json';
-const DARK_MAP_STYLE = 'https://basemaps.cartocdn.com/gl/dark-matter-gl-style/style.json'
-
-const PLACES_PMTILES_URL = 'pmtiles://https://data.source.coop/protomaps/overture/2024-04-16-beta.0/places.pmtiles';
-const PLACES_MAP_STYLE = {
-  'id': 'places',
-  'type': 'circle',
-  'source': 'places',
-  'source-layer': 'places',
-  'filter': [">=", ["get", "confidence"], 0.0],
-  'paint': {
-    'circle-color':
-      ['case',
-        ['==', ['get', '@category'], 'beauty_salon'], '#fb9a99',
-        ['==', ['get', '@category'], 'hotel'], '#33a02c',
-        ['==', ['get', '@category'], 'landmark_and_historical_building'], '#a6cee3',
-        ['==', ['get', '@category'], 'professional_services'], '#fdbf6f',
-        ['==', ['get', '@category'], 'shopping'], '#e31a1c',
-        ['==', ['get', '@category'], 'restaurant'], '#1f78b4',
-        ['==', ['get', '@category'], 'school'], '#ff7f00',
-        ['==', ['get', '@category'], 'accommodation'], '#b2df8a',
-        '#cab2d6'
-      ],
-    'circle-radius': [
-      'interpolate',
-      ['exponential', 2],
-      ['zoom'],
-      0, 1,
-      19, 8
-    ],
-    'circle-stroke-width': [
-      'interpolate',
-      ['exponential', 2],
-      ['zoom'],
-      12, 0,
-      14, 2
-    ],
-    'circle-stroke-color': 'black'
-  }
-};
+import {
+  Map as MapLibreMap,
+  NavigationControl,
+  Source,
+  AttributionControl,
+} from "react-map-gl/maplibre";
+import "maplibre-gl/dist/maplibre-gl.css";
+import * as pmtiles from "pmtiles";
+import maplibregl from "maplibre-gl";
+import { useState, useEffect, useCallback, useRef } from "react";
+import { Layer, GeolocateControl } from "react-map-gl/maplibre";
+import InspectorPanel from "./InspectorPanel";
+import PropTypes from "prop-types";
+import "./InspectorPanel.css";
+import "./CustomControls.css";
+import ThemeSelector from "./ThemeSelector";
 
+const PMTILES_URL =
+  "pmtiles://https://data.source.coop/protomaps/overture/2024-05-16-beta.0/";
 
 const INITIAL_VIEW_STATE = {
-  latitude: 51.0500,
+  latitude: 51.05,
   longitude: 3.7303,
   zoom: 16,
   bearing: 0,
-  pitch: 0
+  pitch: 0,
+};
+
+// this reference must remain constant to avoid re-renders
+const MAP_STYLE = {
+  version: 8,
+  glyphs: "https://demotiles.maplibre.org/font/{fontstack}/{range}.pbf",
+  sources: {},
+  layers: [],
+};
+
+const ThemeSource = ({ name, url }) => {
+  return <Source id={name} type="vector" url={`${url}${name}.pmtiles`} />;
+};
+
+ThemeSource.propTypes = {
+  name: PropTypes.string.isRequired,
+  url: PropTypes.string.isRequired,
+};
+
+const ThemeTypeLayer = ({
+  theme,
+  type,
+  color,
+  point,
+  line,
+  polygon,
+  extrusion,
+}) => {
+  return (
+    <>
+      {point ? (
+        <Layer
```

Comment:
Are there any major benefits to defining the layers in React like this vs. a JSON file?

My concern which styling the map in imperative code is it removes the ability to style the map using Maputnik. @Bonkles mentioned we may have some cartographer resourcing, so it would be good to understand if this will inhibit their workflow.

Reply:
We might be able to pull out JSON to put into Maputnik, but any cartography changes that happen in Maputnik must be ported back into the code manually, so I feel having them work inside the code, JSX or JS etc, is more sustainable.